### ISSUE TITLE
AsyncCallback is not stack-safe

### ISSUE THREAD

**Issue by fthomas** [japgolly/scalajs-react#639]
We've been using `AsyncCallback` as effect type for a basic IndexedDB integration layer and noticed that it blows the stack for deeply nested `flatMap` calls. This happens for example when `traverse`ing a lot of IndexedDB calls for retrieving or setting values.

The issue can be reproduced with this test case:
```scala
import cats.Monad
import cats.effect.IO
import cats.implicits._
import japgolly.scalajs.react.AsyncCallback
import japgolly.scalajs.react.CatsReact.reactAsyncCallbackCatsInstance
import scala.scalajs.js.JavaScriptException
import org.scalatest.funsuite.AnyFunSuite

class StackSafetyTest extends AnyFunSuite {
  def nestedFlatMapsInTailrecLoop[F[_]](implicit F: Monad[F]): F[Int] = {
    val n = 50000
    @scala.annotation.tailrec
    def sum(list: List[F[Int]])(acc: F[Int]): F[Int] =
      list match {
        case Nil          => acc
        case head :: tail => sum(tail)(head.flatMap(h => acc.map(_ + h)))
      }
    sum(List.fill(n)(F.pure(1)))(F.pure(0))
  }

  def nestedFlatMapsInNonTailrecLoop[F[_]](implicit F: Monad[F]): F[Int] = {
    val n = 50000
    def sum(list: List[F[Int]])(acc: F[Int]): F[Int] =
      list match {
        case Nil          => acc
        case head :: tail => head.flatMap(h => sum(tail)(acc.map(_ + h)))
      }
    sum(List.fill(n)(F.pure(1)))(F.pure(0))
  }

  test("AsyncCallback is not stack-safe") {
    intercept[JavaScriptException](nestedFlatMapsInTailrecLoop[AsyncCallback].unsafeToFuture())
  }

  test("AsyncCallback does not trampoline") {
    intercept[JavaScriptException](nestedFlatMapsInNonTailrecLoop[AsyncCallback].unsafeToFuture())
  }

  test("IO is stack-safe") {
    nestedFlatMapsInTailrecLoop[IO].unsafeToFuture().onComplete(println)
  }

  test("IO trampolines") {
    nestedFlatMapsInNonTailrecLoop[IO].unsafeToFuture().onComplete(println)
  }
}
```
Withtout the `intercept[JavaScriptException]` assertions, the test fails with the following exception:
```
 - AsyncCallback is not stack-safe *** FAILED ***
   scala.scalajs.js.JavaScriptException: RangeError: Maximum call stack size exceeded
   at $as_s_util_Try(client-test.js:178621:24)
   at {anonymous}()(client-test.js:8886:22)
   at scala.scalajs.runtime.AnonFunction1.apply(client-test.js:122948:23)
   at {anonymous}()(client-test.js:8950:111)
   at scala.scalajs.runtime.AnonFunction1.apply(client-test.js:122948:23)
   at {anonymous}()(client-test.js:8884:116)
   at scala.scalajs.runtime.AnonFunction1.apply(client-test.js:122948:23)
   at {anonymous}()(client-test.js:8884:116)
   at scala.scalajs.runtime.AnonFunction1.apply(client-test.js:122948:23)
   at {anonymous}()(client-test.js:8905:71)
   ...
```
As comparison, I've aslo included tests of `cats.effect.IO` which demonstrate that it is stack-safe.

This happens with scalajs-react 1.6.0 and Scala.js 0.6.32.

Btw, many thanks for this project, we're enjoying working with scalajs-react a lot!

**Comment by japgolly** [comment]
Hi @fthomas 

> Btw, many thanks for this project, we're enjoying working with scalajs-react a lot!

Let me start with this first: thank you! That's really great to hear. I'm glad you're enjoying it

Now regarding stack safety of both `AsyncCallback` and `Callback(To)`, long story short: its omission thus far was intentional but now that you're bumping up against it (and sorry about that), let's get it sorted.

First things first, I've been unable to max out the stack using your test case. It's surprising because theoretically speaking it's not stack safe - I'm not disputing that at all - I just don't understand why I'm not seeing the results that you are. I bumped up the reps to 1,000,000 and it still passed. When I change it to 10,000,000 reps I get this:

```

<--- Last few GCs --->

[1086656:0x55f937db1490]    16058 ms: Scavenge 2034.2 (2040.3) -> 2034.1 (2045.6) MB, 4.6 / 0.0 ms  (average mu = 0.136, current mu = 0.040) allocation failure 
[1086656:0x55f937db1490]    16065 ms: Scavenge 2037.3 (2045.6) -> 2037.6 (2046.3) MB, 6.2 / 0.0 ms  (average mu = 0.136, current mu = 0.040) allocation failure 
[1086656:0x55f937db1490]    16072 ms: Scavenge 2038.0 (2046.3) -> 2037.6 (2057.3) MB, 6.5 / 0.0 ms  (average mu = 0.136, current mu = 0.040) allocation failure 


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x55f937249999]
Security context: 0x3a7188c80921 <JSObject>
    1: sum$1__p1__sci_List__O__Lcats_Monad__O [0x1abd13cb3e19] [file:///home/golly/projects/public/scalajs-react/test/target/scala-2.13/test-test-fastopt.js:~11579] [pc=0x2db21f502b33](this=0x118e1b949081 <$c_Ljapgolly_scalajs_react_core_StackSafety$ map = 0x788f2f8ab21>,0x1129e6918721 <$c_sci_$colon$colon map = 0x301d91ded8f1>,0x3d5e59087f09 <$c_Ljapgolly_s...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x55f9368e4711 node::Abort() [node]
 2: 0x55f9368e6d05 node::OnFatalError(char const*, char const*) [node]
 3: 0x55f936a5a142 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0x55f936a5a3a8 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0x55f936be46c6  [node]
 6: 0x55f936be4809  [node]
 7: 0x55f936bf6fed v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0x55f936bf7d18 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0x55f936bf91e9 v8::internal::Heap::HandleGCRequest() [node]
10: 0x55f936ba74a6 v8::internal::StackGuard::HandleInterrupts() [node]
11: 0x55f936eeff6b v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [node]
12: 0x55f937249999  [node]
```

"Ok!" I thought. I'll start by making `Callback` stack-safe and I've done that in 8aca622ff27b8ec85660754b92a4bf28b5d79cf6 but I'm still getting the same error. Looking more closely it's saying *JavaScript heap out of memory* which makes sense from a trampolining pov in that we're talking about heap instead of stack (although 10,000,000 seems pretty small).

I don't really know what's going on. I'm hoping to see an out-of-stack error then make it disappear. The fact that all I can get is an OOM regardless of whether I use the original or a trampolined implementation makes me think something else is going wrong.

Any ideas? Would you mind checking out ba74c5552f87e08464987d651fc769eb3de810a0 (tests with orig imp) and 8aca622ff27b8ec85660754b92a4bf28b5d79cf6 (tests + trampoline) seeing if you're have the same experience? Or is there any more info about the environment in which you ran those tests?

Cheers

PR: https://github.com/japgolly/scalajs-react/pull/640

**Comment by fthomas** [comment]
Thanks for the swift response, @japgolly! I've no idea why you can't reproduce the stack issue. To make sure that it isn't caused by any particularity in my local environment, I've set up a new [project](https://github.com/fthomas/asynccallback-stack-safety) that reproduces this on Travis CI: https://travis-ci.com/fthomas/asynccallback-stack-safety/builds/147509608. The exception in the Travis build is the same as the one I posted before.

I'll try now the commits you have linked to.

Btw, for very large values of `n` I also observe OOMEs for both `AsyncCallback` and `cats.effect.IO`.

**Comment by fthomas** [comment]
I just tested both commits and found that the exceptions can be observed if `.toCallback` or `.unsafeToFuture()` is added to the tests (https://github.com/japgolly/scalajs-react/pull/640/files#r375185064). Unfortunately I see this exception with both commits.

**Comment by japgolly** [comment]
#640 has been merged into master now and will go out in v1.6.1

Thanks for your help @fthomas !

**Comment by fthomas** [comment]
NP. I thank you for the quick fix!
